web/src/typing.ts

```diff
@@ -71,13 +92,33 @@ function send_typing_notification_based_on_message_type(
     to: Recipient,
     operation: "start" | "stop",
 ): void {
+    if (to.notification_event_type !== "typing") {
+        return;
+    }
     if (to.message_type === "direct" && user_settings.send_private_typing_notifications) {
         send_direct_message_typing_notification(to.ids, operation);
     } else if (to.message_type === "stream" && user_settings.send_stream_typing_notifications) {
         send_stream_typing_notification(to.stream_id, to.topic, operation);
     }
 }
 
+function send_typing_notifications_for_message_edit(
+    message_id: number,
+    operation: "start" | "stop",
+): void {
+    if (message_edit_typing_notifications_enabled(message_id)) {
+        send_message_edit_typing_notification_ajax(message_id, operation);
+    }
+}
+
+function message_edit_typing_notifications_enabled(message_id: number): boolean {
+    const message = message_lists.current?.get(message_id);
+    if (message?.type === "stream") {
```

Comment:
Should you be using `assert` statements or `!` here? I think if you look up the message in `message_store` rather than `message_lists.current`, we should be able to avoid thinking about whether `current` is undefined. And that should only fail on invalid message IDs or those that have been deleted from the message store, and I would hope that this code path would not run for a deleted message. (You might need some code to do cleanup in the "delete message" handler code paths).

And certainly one should be able to assume that a message object has a `type`  field.